File: src/breslin/clientside/game/application.cpp
//header
#include "application.h"

//log
#include "../tdreamsock/dreamSockLog.h"

//shape
#include "../shape/shape.h"

//math
#include "../../math/vector3D.h"

//byteBuffer
#include "../byteBuffer/byteBuffer.h"

//network
#include "../network/network.h"

//command
#include "../command/command.h"

//game
#include "game.h"


/***************************************
*			          CONSTRUCTORS
***************************************/
Application::Application(const char* serverIP, int serverPort)
{
	StartLog();

	// network
	mNetwork = new Network(this,serverIP,serverPort);

	//time
	mFrameTime		 = 0.0f;

	//time
	mRenderTime = 0;

	//keys
	mKeyUp = 1;
	mKeyDown = 2;
	mKeyLeft = 4;
	mKeyRight = 8;
	mKeyCounterClockwise = 16;
	mKeyClockwise = 32;

	//initilize
	mInitializeGui = false;
	mJoinGame    = false;
	mPlayingGame = false;

	//game
	mGame = new Game(this,serverIP,serverPort);

	//input
	mKeyCurrent = 0;
	mKeyLast;
	mMillisecondsCurrent = 0;
	mMillisecondsLast = 0;
}

Application::~Application()
{
	mNetwork->close();
	delete mNetwork;
}

/*********************************
		START/LOOP/END
**********************************/
void Application::run()
{
	while(true)
    {
		//input
		processInput();
		
		//network
		runNetwork(getRenderTime() * 1000.0f);
		
		//move objects
		mGame->run();

		//draw
		if (!runGraphics())
		{
			break;
		}
	}
}

void Application::shutdown()
{
	mNetwork->sendDisconnect();
}


void Application::readServerTick(ByteBuffer* byteBuffer)
{
	// Skip sequences
	byteBuffer->ReadShort();

	while (byteBuffer->getReadCount() <= byteBuffer->GetSize())
	{
		//mDetailsPanel->setParamValue(11, Ogre::StringConverter::toString(byteBuffer->GetSize()));

		int id = byteBuffer->ReadByte();

		Shape* shape = NULL;
		shape = mGame->getShape(id);

		if (shape)
		{
			shape->readDeltaMoveCommand(byteBuffer);
			shape->processTick();
		}
	}
}

/*********************************
		NETWORK
**********************************/
void Application::runNetwork(float msec)
{
	static float time = 0.0f;
	time += msec;

	mNetwork->readPackets();
	
	// Framerate is too high
	if(time > (1000 / 60))
	{
		sendCommand();
		mFrameTime = time / 1000.0f;
		time = 0;
	}
}

/**********************************************
*
*		OGRE_SPECIFIC
*
**************************************************



/*********************************
*		TIME
***********************************/
float Application::getRenderTime()
{
	//LogString("r:%f",mGraphicsOgre->mRenderTime);
	return mRenderTime;
}


/*********************************
		GRAPHICS
**********************************/
void Application::createScene(void)
{
    mSceneMgr->setAmbientLight(Ogre::ColourValue(0.75, 0.75, 0.75));

    Ogre::Light* pointLight = mSceneMgr->createLight("pointLight");
    pointLight->setType(Ogre::Light::LT_POINT);
    pointLight->setPosition(Ogre::Vector3(250, 150, 250));
    pointLight->setDiffuseColour(Ogre::ColourValue::White);
    pointLight->setSpecularColour(Ogre::ColourValue::White);
}

bool Application::frameRenderingQueued(const Ogre::FrameEvent& evt)
{
	mRenderTime = evt.timeSinceLastFrame;

    bool ret = BaseApplication::frameRenderingQueued(evt);
	
	return ret;
}

bool Application::runGraphics()
{
	//Pump messages in all registered RenderWindow windows
	WindowEventUtilities::messagePump();
	if (!mRoot->renderOneFrame())
	{
		initializeGui();
		return false;
	}
	else
	{
		initializeGui();
		return true;
	}


}

/*********************************
		GUI
**********************************/
void Application::initializeGui()
{
	if (mInitializeGui == true)
	{
		return;
	}
	else
	{
		loadJoinScreen();
		mInitializeGui = true;
	}
}

void Application::loadJoinScreen()
{
	mJoinButton = mTrayMgr->createButton(OgreBites::TL_CENTER, "mJoinButton", "Join Game");
	mTrayMgr->moveWidgetToTray(mJoinButton,OgreBites::TL_CENTER);
	mTrayMgr->showCursor();
}

void Application::hideGui()
{
	hideJoinScreen();
	mTrayMgr->hideCursor();
}

void Application::hideJoinScreen()
{
	mTrayMgr->removeWidgetFromTray(mJoinButton);
    mJoinButton->hide();
}


/***************************************
*			INPUT
******************************************/

void Application::buttonHit(OgreBites::Button *button)
{
	//JOIN
	if (button == mJoinButton)
	{
		mJoinGame = true;
		if (mJoinGame && !mPlayingGame)
		{
			mNetwork->sendConnect();
			mPlayingGame = true;
		}
		hideGui();
	}
}

bool Application::mouseMoved( const OIS::MouseEvent &arg )
{
    if (mTrayMgr->injectMouseMove(arg))
	{
		return true;
	}
	if (mPlayingGame)
	{
		mCameraMan->injectMouseMove(arg);
	}
    return true;
}

void Application::processInput()
{
	mKeyCurrent = 0;
    
	if (mKeyboard->isKeyDown(OIS::KC_I)) // Forward
    {
		mKeyCurrent |= mKeyUp;
    }

    if (mKeyboard->isKeyDown(OIS::KC_K)) // Backward
    {
		mKeyCurrent |= mKeyDown;
    }

	if (mKeyboard->isKeyDown(OIS::KC_J)) // Left
    {
		mKeyCurrent |= mKeyLeft;
    }

    if (mKeyboard->isKeyDown(OIS::KC_L)) // Right
    {
		mKeyCurrent |= mKeyRight;
    }
    
	if (mKeyboard->isKeyDown(OIS::KC_Z)) // Rotate -Yaw(counter-clockwise)
    {
		mKeyCurrent |= mKeyCounterClockwise;
    }

    if (mKeyboard->isKeyDown(OIS::KC_X)) // Right + Yaw(clockwise)
    {
		mKeyCurrent |= mKeyClockwise;
    }
	mMillisecondsCurrent = (int) (mFrameTime * 1000);
}


/***************************************************
*			COMMAND
***************************************************/

void Application::sendCommand(void)
{
	//create byteBuffer
	ByteBuffer* byteBuffer = new ByteBuffer();

	//WRITE: type
	byteBuffer->WriteByte(mMessageFrame);					
	
	//WRITE: sequence
	byteBuffer->WriteShort(mNetwork->mOutgoingSequence);

	// Build delta-compressed move command
	int flags = 0;

	// Check what needs to be updated
	if(mKeyLast != mKeyCurrent)
	{
		flags |= mCommandKey;
	}

	if(mMillisecondsLast != mMillisecondsCurrent)
	{
		flags |= mCommandMilliseconds;
	}
	
	// Add to the message
	byteBuffer->WriteByte(flags);

	if(flags & mCommandKey)
	{
		//WRITE: key
		byteBuffer->WriteByte(mKeyCurrent);
	}

	if(flags & mCommandMilliseconds)
	{
		//WRITE: milliseconds
		byteBuffer->WriteByte(mMillisecondsCurrent);
	}
	
	//set 'last' commands for diff
	mKeyLast = mKeyCurrent;
	mMillisecondsLast = mMillisecondsCurrent;

	// Send the packet
	mNetwork->send(byteBuffer);
}

File: src/breslin/clientside/game/Application.java
package breslin.clientside.game;

/***************************************
*   		INCLUDES
***************************************/

//parent
import com.jme3.app.SimpleApplication;

//game
import breslin.clientside.game.GameMonkey;

//network
import breslin.clientside.network.Network;

//standard library
import com.jme3.math.Vector3f;


//parent
import com.jme3.app.SimpleApplication;


//lwgl
import org.lwjgl.LWJGLException;
import org.lwjgl.input.Keyboard;
import org.lwjgl.input.Mouse;
import org.lwjgl.opengl.Display;
import org.lwjgl.opengl.DisplayMode;

//input
import com.jme3.input.KeyInput;
import com.jme3.input.controls.ActionListener;
import com.jme3.input.controls.KeyTrigger;

//light
import com.jme3.light.DirectionalLight;


import java.nio.ByteBuffer;


import java.io.UnsupportedEncodingException;

/***************************************
*   		FORWARD DECLARATIONS
***************************************/

public class Application extends SimpleApplication
{

public Application(byte[] serverIP, int serverPort)
{
//from game
	//StartLog();

	try
	{
		String serverIPString = new String(serverIP, "UTF8");
		System.out.println("serverIP:" + serverIPString + " serverPort:" + serverPort);

	}
	catch (UnsupportedEncodingException e)
	{
	    e.printStackTrace();
	}

	// network
	mNetwork = new Network(this,serverIP,serverPort);

	//time
	//mTime = new Time();
	mFrameTime		 = 0.0f;
	mOldTime         = 0;


	//game
	mGameMonkey = gameMonkey;

	//keys
	mKeyUp = 1;
	mKeyDown = 2;
	mKeyLeft = 4;
	mKeyRight = 8;
	mKeyCounterClockwise = 16;
	mKeyClockwise = 32;

	//initilize
	mInitializeGui = false;
	mJoinGame    = false;
	mPlayingGame = false;

	start();

	//let their be light
	DirectionalLight directionalLight = new DirectionalLight();
	directionalLight.setDirection(new Vector3f(-0.1f, -1f, -1).normalizeLocal());
    getRootNode().addLight(directionalLight);

mRenderTime = 0;

}


/**************************************
*			MAIN
**************************************/
public static void main(String[] args)
{
	byte[] theByteArray;
	if (args.length > 0)
	{
		theByteArray = args[0].getBytes();
	}
	else
	{
		String s = "192.168.1.104";
		theByteArray = s.getBytes();
	}
	Application application = new Application(theByteArray,30004);
}

/***************************************
*   		MEMBER VARIABLES
***************************************/

//initialize
boolean mInitializeGui;
boolean mJoinGame;
boolean mPlayingGame;

//game
GameMonkey mGameMonkey;

//Network
public Network     mNetwork;

//gui
//OgreBites::Button* mJoinButton;

//keys
int mKeyUp;
int mKeyDown;
int mKeyLeft;
int mKeyRight;
int mKeySpace;

int mKeyCounterClockwise;
int mKeyClockwise;

public float mFrameTime;
public float mRunNetworkTime;
public int   mOldTime;

//render time
public float mRenderTime;

/***************************************
*			          METHODS
***************************************/

public void simpleInitApp()
{

}

public void update()
{
	//input
	processInput();

	//network
	mGameMonkey.runNetwork(getRenderTime() * 1000.0f);

	//move objects
	mGameMonkey.interpolateTick();

	//super and other stuff
	super.update();
}

//graphics
/* This is the update loop */
public void simpleUpdate(float tpf)
{
    mGameMonkey.mRenderTime = tpf;
}

void        createScene          ()
{

}

//boolean        frameRenderingQueued (const Ogre::FrameEvent& evt);

public boolean                runGraphics          ()
{
	super.update();
	return true;
}

//gui
void hideGui()
{

}

void loadJoinScreen()
{

}

void hideJoinScreen()
{

}
void unloadOtherScreens()
{

}
void initializeGui()
{

}

//input
void processInput()
{
	mGameMonkey.mNetwork.mCommandToServer.mKey = 0;

	if (Keyboard.isKeyDown(Keyboard.KEY_B))
	{
		if (!mPlayingGame)
		{
			mGameMonkey.mNetwork.sendConnect();
			mPlayingGame = true;

			//Set Camera to position and to lookat avatar at 0,0,0(this should be same as ogre! if not fix it)
			Vector3f startCamPosition = new Vector3f(0, 20, 20);
			Vector3f lookAtVector     = new Vector3f(0,0,0);
			Vector3f worldDirection   = new Vector3f(0,1,0);
			cam.setLocation(startCamPosition);
			cam.lookAt(lookAtVector,worldDirection);
		}
		else
		{
			System.out.println("slow your roll you already started");
		}
	}

	//move
	if (Keyboard.isKeyDown(Keyboard.KEY_I))
	{
		mGameMonkey.mNetwork.mCommandToServer.mKey |= mKeyUp;
	}


	if (Keyboard.isKeyDown(Keyboard.KEY_K))
	{
		mGameMonkey.mNetwork.mCommandToServer.mKey |= mKeyDown;
	}


	if (Keyboard.isKeyDown(Keyboard.KEY_J))
	{
		mGameMonkey.mNetwork.mCommandToServer.mKey |= mKeyLeft;
	}


	if (Keyboard.isKeyDown(Keyboard.KEY_L))
	{
		mGameMonkey.mNetwork.mCommandToServer.mKey |= mKeyRight;
	}

	//rotation
	if (Keyboard.isKeyDown(Keyboard.KEY_G))
	{
		mGameMonkey.mNetwork.mCommandToServer.mKey |= mKeyCounterClockwise;
	}

	if (Keyboard.isKeyDown(Keyboard.KEY_H))
	{
		mGameMonkey.mNetwork.mCommandToServer.mKey |= mKeyClockwise;
	}

	mGameMonkey.mNetwork.mCommandToServer.mMilliseconds = (byte) (mGameMonkey.mFrameTime * 1000);
}


public Vector3f getCameraLocation()
{
	return cam.getLocation();
}

public float getRenderTime()
{
//	System.out.println("r:" + mRenderTime);
	return mRenderTime;
}

/***************************************
*   		TICKS
***************************************/
public void readServerTick           (ByteBuffer byteBuffer)
{
	// Skip sequences
	//short sequence = byteBuffer.get();

	//WRITE: sequence
	//byteBuffer.putShort(mOutgoingSequence);  //sequence
	byte one = byteBuffer.get(1);
	byte two = byteBuffer.get(2);
	byteBuffer.put(1,two);
	byteBuffer.put(2,one);
	byteBuffer.position(1);
	short sequence = byteBuffer.getShort();
//	System.out.println("sequence:" + sequence);





	while (byteBuffer.hasRemaining())
	{

		//mDetailsPanel->setParamValue(11, Ogre::StringConverter::toString(byteBuffer->GetSize()));

		int id = byteBuffer.get();
		//System.out.println("id:" + id);

		Shape shape = null;
		shape = getShape(id);

		if (shape != null)
		{
			//System.out.println("got shape");
			shape.readDeltaMoveCommand(byteBuffer);
		}

	}

}


public void interpolateTick()
{



}


/***************************************
*   		NETWORK
***************************************/
public void runNetwork    (float msec)
{
	mRunNetworkTime += msec;

	mNetwork.readPackets();

	// Framerate is too high
	if(mRunNetworkTime > (1000 / 60))
	{
		mNetwork.sendCommand();
		mFrameTime = mRunNetworkTime / 1000.0f;
		mRunNetworkTime = 0;
	}
}

public void shutdown()
{

}

};

